### ISSUE TITLE
Neither decompileContext nor clearContext actually blocks until the decompile is finished

### ISSUE THREAD

**Issue by TwilightFlower** [FabricMC/intellij-fernflower#12]
I'm not exactly sure what causes this, but it doesn't seem like intended behavior. Right now I'm using a workaround of blocking until the first closeArchive call.

**Comment by Chocohead** [comment]
This will be a consequence of what the particular `IResultSaver` you're using will be doing.

If we look at `Fernflower#decompileContext`:
https://github.com/FabricMC/intellij-fernflower/blob/6f54090349af9cb1b4663765cf4bd2e195dd53c0/src/org/jetbrains/java/decompiler/main/Fernflower.java#L80-L88

The final call is to `StructContext#saveContext`. That method will save all the `ContextUnits` which are being decompiled, the ones which aren't represent libraries which are only there for building a classpath up:
https://github.com/FabricMC/intellij-fernflower/blob/6f54090349af9cb1b4663765cf4bd2e195dd53c0/src/org/jetbrains/java/decompiler/struct/StructContext.java#L54-L60

`ContextUnit#save` is a pretty long method, so for the sake of simplicity we'll just look at what happens for zips/jars:
https://github.com/FabricMC/intellij-fernflower/blob/6f54090349af9cb1b4663765cf4bd2e195dd53c0/src/org/jetbrains/java/decompiler/struct/ContextUnit.java#L117-L170

First it creates the directory structure for exporting the produced java files, and copies over any resources which were also present in the `ContextUnit`. Next, each class has a `Future` made which represents the decompiled output (if it has any). The executor given these `Future`s to run is then shutdown to encourage them to all finish and each `Future` joined in turn to make sure they _have_ completed. At this point FernFlower has done its job and produced all the output it has to offer for the particular `ContextUnit`. The final step is to notify the `IResultSaver` that the archive opened (if there was one) can now be closed as there is nothing else for it.

You should notice from this that at no point has the thread which called `Fernflower#decompileContext` left any of its own tasks running. It has certainly utilised them, but made sure they've all finished before it returns. To explain the behaviour you're seeing therefore something else has allowed the main thread to continue without actually having finished the work - the `IResultSaver` being given the output. Both [Loom's](https://github.com/FabricMC/fabric-loom/blob/679026ef31d619409187a564bc0623891f5979e6/src/main/java/net/fabricmc/loom/decompilers/fernflower/ThreadSafeResultSaver.java#L128-L156) and [ForgeFlower's](https://github.com/Chocohead/ForgedFlower/blob/d6b8cf1f90846da462ac5c538245cc4b7e6b5689/src/org/jetbrains/java/decompiler/main/decompiler/ThreadSafeResultSaver.java#L127-L158) implementations of `IResultSaver#closeArchive` will block until any extra tasks they've created have finished, which is effectively the best that can be done, if you've written your own implementation I'd double check you've got something like that.

**Comment by TwilightFlower** [comment]
Yeah, that's what I ended up doing -- however, this doesn't seem like behavior that should just happen without being documented anywhere, as it does not seem at all intended based on the program structure, and the hook in IResultSaver is non-obvious due to naming.
